freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/PasswordService.java

```diff
@@ -41,48 +54,103 @@
+                stacks.forEach(stack -> LOGGER.debug("Found matching stack for environment {}", stack.getEnvironmentCrn()));
+            }
+        } else {
+            stacks = allStacks;
+        }
 
         if (stacks.isEmpty()) {
-            throw new IllegalArgumentException("No stacks found for accountId " + accountId);
+            LOGGER.warn("No stacks found for accountId {}", accountId);
+            throw new NotFoundException("No stacks found for accountId " + accountId);
+        }
+
+        Future<List<SetPasswordRequest>> requests = asyncTaskExecutor.submit(() -> triggerSetPasswords(accountId, userCrn, password, stacks));
+
+        SetPasswordResponse response = SetPasswordResponse.running(operationId, startTime);
+        passwordOperationStatusCache.put(operationId, response);
+
+        asyncTaskExecutor.submit(() -> waitSetPasswords(operationId, accountId, userCrn, requests));
+
+        return response;
+    }
+
+    public SetPasswordResponse getSetPasswordStatus(String operationId) {
+        SetPasswordResponse response = passwordOperationStatusCache.getIfPresent(operationId);
+        if (response == null) {
+            throw new NotFoundException("Status not found for operation " + operationId);
         }
+        return response;
+    }
+
+    private List<SetPasswordRequest> triggerSetPasswords(String accountId, String userCrn, String password, List<Stack> stacks) {
+        String userId = getUserIdFromUserCrn(userCrn);
 
         List<SetPasswordRequest> requests = new ArrayList<>();
         for (Stack stack : stacks) {
+            requests.add(triggerSetPassword(stack, stack.getEnvironmentCrn(), userId, password));
+        }
+        return requests;
+    }
 
-            // Filter based on provided envs
-            if (envs != null && envs.size() > 0) {
-                if (envs.contains(stack.getEnvironmentCrn())) {
-                    LOGGER.debug("Env list provided, setting password for env :{}", stack.getEnvironmentCrn());
-                    requests.add(triggerSetPassword(stack, stack.getEnvironmentCrn(), userId, password));
-                }
-            } else {
-                requests.add(triggerSetPassword(stack, stack.getEnvironmentCrn(), userId, password));
-            }
+    private void waitSetPasswords(String operationId, String accountId, String userCrn, Future<List<SetPasswordRequest>> requestsFuture) {
+        SetPasswordResponse response = passwordOperationStatusCache.getIfPresent(operationId);
+        long startTime = response != null ? response.getStartTime() : System.currentTimeMillis();
+
+        List<SetPasswordRequest> requests;
+        try {
+            requests = requestsFuture.get();
+        } catch (InterruptedException | ExecutionException e) {
+            passwordOperationStatusCache.put(operationId, SetPasswordResponse.failed(operationId, e.getLocalizedMessage(),
+                    startTime, System.currentTimeMillis()));
+            return;
```

Comment:
not too elegant returning in a catch part. everything should be in the try part which depends on it

Reply:
done